Put the Compiler standard option in the Language mixin

dcbaker opened the pull request:
This has a nice effect on code duplication, plus ensures that all compilers expose the {lang}_std option, even if the only option is none. This prevents cases where a project says "I need gnu99" but the compiler doesn't support it, and meson acts like it does.

This also cleans up the options, as some use (incorrectly) `c_std`, when they should use just `std`

dcbaker commented:
Hmmm, now that I think about it, clang-cl has the `/clang` option for proxying options to clang itself. We might be able to do `/clang:-std=c99`, etc. Which would be better.

Let's hold off on merging this until I have a chance to look at that.

nirbheek commented on mesonbuild/compilers/c.py:
```diff
@@ -420,6 +420,11 @@ def get_options(self) -> 'OptionDictType':
+            mlog.warning(
+                'cl.exe does not actually support gnu standards, and meson '
+                'will instead demote to the nearest ISO C standard. This '
+                'may cause compilation to fail.', once=True)
```
We should not add this warning till we have a replacement for this workflow.

Please, this is really really bad. We should **never** be spammy about warnings. It makes `--fatal-meson-warnings` unusable by projects and makes it harder for us to deprecate things because people just ignore warnings.

dcbaker replied:
I've replaced the warning with a log() message instead, as the message seems useful to me (all of my projects that use gnu99 will fail to compile with c99, so giving me *something* is nice.) I've also replaced a similar warning in the IntelCl compiler with a log() message instead.

I'll get to adding a fallback mechanism later, but for now is this sufficient to land?

dcbaker commented:
The LGTM is a lie, i've moved some things around, so it thinks i've "fixed" things and "broken" things, but I've just moved existing warnings from here to there.

jpakkane commented:
```
  File "D:\a\1\s\mesonbuild\backend\ninjabackend.py", line 843, in generate_target
    elem = self.generate_link(target, outname, obj_list, linker, pch_objects, stdlib_args=stdlib_args)
  File "D:\a\1\s\mesonbuild\backend\ninjabackend.py", line 2789, in generate_link
    commands += linker.get_option_link_args(self.environment.coredata.compiler_options[target.for_machine][linker.get_language()])
  File "D:\a\1\s\mesonbuild\compilers\c.py", line 411, in get_option_link_args
    libs = options['winlibs'].value.copy()
KeyError: 'winlibs'
```
